Fix endianness of string literals

mhillenbrand opened the pull request:
To get correct and consistent encoding and decoding of string literals
on big-endian platforms, use spvtools::utils::MakeString and MakeVector
(or wrapper functions) consistently for handling string literals.

- add variant of MakeVector that encodes a string literal into an
  existing vector of words
- add variants of MakeString
- add a wrapper spvDecodeLiteralStringOperand in source/
- fix wrapper Operand::AsString to use MakeString (source/opt)
- remove Operand::AsCString as broken and unused
- add a variant of GetOperandAs for string literals (source/val)
... and apply those wrappers throughout the code.

Extend test cases
- StringLiterals.RoundTripLiteralsTest,
- StringLiterals.RoundTripSpecialCaseLiteralsTest, and
- BinaryParseTest.InstructionWithStringOperand

to attempt decoding both with the host's own byte order and with flipped words, that is, test decoding SPIR-V binaries in both little-endian and big-endian orders. The cross-endian test cases fail without the fixes in this PR.

Fixes  #149

Tested on x86-64 and s390x, there are no regressions in the testsuite.

greg-lunarg commented:
Let me know when this is merged and I will merge https://github.com/KhronosGroup/glslang/pull/2800

dneto0 commented on source/util/string_utils.h:
```diff
@@ -44,9 +46,9 @@ std::string CardinalToOrdinal(size_t cardinal);
-  std::vector<uint32_t> result;
+// Encodes a string as a sequence of words, using the SPIR-V encoding, appending
+// to an existing vector.
+inline void AppendToVector(std::string input, std::vector<uint32_t>& result) {
```
Please avoid non-const references.
This codebase has avoided them (except for I/O streams).

mhillenbrand commented:
I have added two commits that extend the test cases StringLiterals.RoundTripLiteralsTest,
StringLiterals.RoundTripSpecialCaseLiteralsTest, and BinaryParseTest.InstructionWithStringOperand to test decoding from both little-endian and big-endian byte order. The tests encode into the host's own byte order as usual and then do / do not flip byte order before attempting decode, as many other test cases already do.

As expected, the "cross-endian" test cases would fail without the fixes in this patch because the decoding step would always assume the host's byte order.

I have updated this PR's description to reflect this extension.

mhillenbrand commented:
Steven @s-perron, you have added encoding and decoding of string literals in source/util/string_utils.[h|cpp]. Could you please take a look at this PR extending your code?

s-perron commented on source/util/string_utils.h:
```diff
@@ -56,23 +58,33 @@ inline std::vector<uint32_t> MakeVector(std::string input) {
-  for (uint32_t word : words) {
+  for (InputIt pos = first; pos != last; ++pos) {
+    uint32_t word = *pos;
     for (int byte_index = 0; byte_index < 4; byte_index++) {
```
I know you did not do this, but we should probably not be use "4" as a literal here.  We use 4 because we are assuming that "*pos" is 4 bytes.  We should provide a named constant for this, and add a static assert.

mhillenbrand replied:
Agreed, good catch.

s-perron commented on source/util/string_utils.h:
```diff
@@ -56,23 +58,33 @@ inline std::vector<uint32_t> MakeVector(std::string input) {
+// assertFoundTerminatingNull is passed as false).
+template <class InputIt>
+inline std::string MakeString(InputIt first, InputIt last,
+                              bool assertFoundTerminatingNull = true) {
```
We follow the Google style guide (https://google.github.io/styleguide/cppguide.html#Variable_Names).  The variable name should be `assert_fount_terminating_null`.  I know not everything follows it.  Some code predates us using that style.  New code should try to fix it up.

s-perron reviewed (changes requested):
LGTM.  There is just a few variable names that need to be changed to have the correct format. Sorry I did not spot this PR earlier.

s-perron commented:
@mhillenbrand The CI runs use the latest from all of the dependencies, so you will need to rebase on top of the latest master to get the tests to pass.

mhillenbrand commented:
@s-perron Thanks for starting the test runs. I have rebased and checked the build + local testsuite with updated dependencies (git-sync-deps)

s-perron commented:
Look like there might be a memory leak somewhere.  The asan build it failing.  Can you look into that.  If you cannot see anything or you do not believe it is related to your change, then let me know and I will look into it.

mhillenbrand commented:
I will take a look today. I cannot access the logs of the failing CI run (the link "Details" leads nowhere) yet I should be able to reproduce that locally with the scripts in kokoro/.

s-perron commented:
I just noticed another PR failed the asan build in same way.  I'll double check it, but I might merge this anyway.

mhillenbrand commented:
Thanks for merging!

I reproduced the ASAN runs as close to the CI as I could -- using `./kokoro/linux-clang-asan/build.sh` and slightly adjusting `./kokoro/scripts/linux/build.sh` -- yet the test run completed successfully. fwiw, in an early run LeakSanitizer bailed out because I was missing options to docker (`--cap-add SYS_PTRACE`).
